servers/src/main/java/tachyon/worker/block/evictor/EvictorUtils.java

```diff
@@ -42,46 +40,68 @@ public static boolean legalCascadingPlan(long bytesToBeAvailable, EvictionPlan p
 
-    List<Long> blockIds = new ArrayList<Long>(plan.toEvict().size() + plan.toMove().size());
-    blockIds.addAll(plan.toEvict());
-    for (Pair<Long, BlockStoreLocation> move : plan.toMove()) {
-      blockIds.add(move.getFirst());
+    // Map from dir to a pair of bytes to be available in this dir and bytes to move into this dir
+    // after the plan taking action
+    Map<StorageDir, Pair<Long, Long>> spaceInfoInDir = new HashMap<StorageDir, Pair<Long, Long>>();
+
+    for (long blockId : plan.toEvict()) {
+      BlockMeta block = metaManager.getBlockMeta(blockId);
+      StorageDir dir = block.getParentDir();
+      if (spaceInfoInDir.containsKey(dir)) {
+        Pair<Long, Long> spaceInfo = spaceInfoInDir.get(dir);
+        spaceInfo.setFirst(spaceInfo.getFirst() + block.getBlockSize());
+        spaceInfoInDir.put(dir, spaceInfo);
+      } else {
+        spaceInfoInDir.put(dir, new Pair<Long, Long>(
+            dir.getAvailableBytes() + block.getBlockSize(), 0L));
+      }
     }
 
-    for (long blockId : blockIds) {
+    for (Pair<Long, BlockStoreLocation> move : plan.toMove()) {
+      long blockId = move.getFirst();
       BlockMeta block = metaManager.getBlockMeta(blockId);
-      BlockStoreLocation blockDir = block.getBlockLocation();
       long blockSize = block.getBlockSize();
-      int tierAlias = blockDir.tierAlias();
-      if (bytesToBeAvailableInTier.containsKey(tierAlias)) {
-        bytesToBeAvailableInTier
-            .put(tierAlias, bytesToBeAvailableInTier.get(tierAlias) + blockSize);
+      StorageDir srcDir = block.getParentDir();
+      StorageDir destDir = metaManager.getDir(move.getSecond());
+
+      if (spaceInfoInDir.containsKey(srcDir)) {
+        Pair<Long, Long> spaceInfo = spaceInfoInDir.get(srcDir);
+        spaceInfo.setFirst(spaceInfo.getFirst() + blockSize);
+        spaceInfoInDir.put(srcDir, spaceInfo);
       } else {
-        tierAliases.add(tierAlias);
-        bytesToBeAvailableInTier
-            .put(tierAlias, metaManager.getAvailableBytes(blockDir) + blockSize);
+        spaceInfoInDir
+            .put(srcDir, new Pair<Long, Long>(srcDir.getAvailableBytes() + blockSize, 0L));
+      }
+
+      if (spaceInfoInDir.containsKey(destDir)) {
+        Pair<Long, Long> spaceInfo = spaceInfoInDir.get(destDir);
+        spaceInfo.setSecond(spaceInfo.getSecond() + blockSize);
+        spaceInfoInDir.put(destDir, spaceInfo);
+      } else {
+        spaceInfoInDir.put(destDir, new Pair<Long, Long>(destDir.getAvailableBytes(), blockSize));
       }
     }
 
-    // upper to lower tier
-    Collections.sort(tierAliases);
-    int currentTierAlias = tierAliases.get(0);
-    // first tier to free space from needs to have bytesToBeAvailable after eviction
-    boolean isLegal = bytesToBeAvailableInTier.get(currentTierAlias) >= bytesToBeAvailable;
-    for (int nextTierAlias : tierAliases.subList(1, tierAliases.size())) {
-      if (!isLegal) {
-        break;
+    StorageDir firstTierDir = null;
```

Comment:
Would it be possible to avoid assuming only one dir will be affected by eviction in the first tier (for example you could just make sure the max of the free space of dirs in the first tier is > bytesToBeAvailable)?